evalml/model_understanding/_partial_dependence.py

```diff
@@ -0,0 +1,251 @@
+"""Partial dependence implementation.
+
+Borrows from sklearn "brute" calculation but with our
+own modification to better handle mixed data types in the grid
+as well as EvalML pipelines.
+"""
+import numpy as np
+import pandas as pd
+from scipy.stats.mstats import mquantiles
+
+from evalml.problem_types import is_regression
+
+
+def _range_for_dates(X_dt, percentiles, grid_resolution):
+    """Compute the range of values used in partial dependence for datetime features.
+
+    Interpolate between the percentiles of the dates converted to unix
+    timestamps.
+
+    Args:
+        X_dt (pd.DataFrame): Datetime features in original data. We currently
+            only support X_dt having a single column.
+        percentiles (tuple float): Percentiles to interpolate between.
+        grid_resolution (int): Number of points in range.
+
+    Returns:
+        pd.Series: Range of dates between percentiles.
+    """
+    timestamps = np.array(
+        [X_dt - pd.Timestamp("1970-01-01")] // np.timedelta64(1, "s")
+    ).reshape(-1, 1)
+    timestamps = pd.DataFrame(timestamps)
+    grid, values = _grid_from_X(
+        timestamps,
+        percentiles=percentiles,
+        grid_resolution=grid_resolution,
+        custom_range={},
+    )
+    grid_dates = pd.to_datetime(pd.Series(grid.squeeze()), unit="s")
+    return grid_dates
+
+
+def _grid_from_X(X, percentiles, grid_resolution, custom_range):
+    """Create cartesian product of all the columns of input dataframe X.
+
+    Args:
+        X (pd.DataFrame): Input data
+        percentiles (tuple float): Percentiles to use as endpoints of the grid
+            for each feature.
+        grid_resolution (int): How many points to interpolate between percentiles.
+        custom_range (dict[str, np.ndarray]): Mapping from column name in X to
+            range of values to use in partial dependence. If custom_range is specified,
+            the percentile + interpolation procedure is skipped and the values in custom_range
+            are used.
+
+    Returns:
+        pd.DataFrame: Cartesian product of input columns of X.
+    """
+    values = []
+    for feature in X.columns:
+        if feature in custom_range:
+            # Use values in the custom range
+            feature_range = custom_range[feature]
+            if not isinstance(feature_range, (np.ndarray, pd.Series)):
+                feature_range = np.array(feature_range)
+            if feature_range.ndim != 1:
+                raise ValueError(
```

Comment:
I'm ok if this isn't covered. It's impossible to trigger this as a user because `custom_range` is not a public parameter. But I'd like to keep this check in case we refactor this in the future. Helped catch a couple of bugs during development.